common/dynamicconfig/mutable_ephemeral_client.go

```diff
@@ -0,0 +1,171 @@
+// Mutations are batched to the Update function. Data written to objects of this
+// type is not made durable.
+type MutableEphemeralClient struct {
+	*basicClient
+	mu sync.Mutex
+}
+
+// NewMutableEphemeralClient constructs a new MutableEphemeralClient with an
+// empty internal store.
+func NewMutableEphemeralClient(mutations ...Mutation) *MutableEphemeralClient {
+	c := &MutableEphemeralClient{basicClient: newBasicClient()}
+	c.Update(mutations...)
+	return c
+}
+
+// MutationConstraint expresses a configuration constraint (namespace, queue
+// name, or queue type) that is an optional part of a Mutation.
+type MutationConstraint func(map[string]interface{})
+
+// Mutation expresses a modification to a MutableEphemeralClient's internal
+// value store. Mutation functions are applied with
+// MutableEphemeralClient.Update.
+type Mutation func(configValueMap)
+
+// ForNamespaces builds an exact-match namespace MutationConstraint.
+func ForNamespace(ns string) MutationConstraint {
+	return func(m map[string]interface{}) {
+		m[Namespace.String()] = ns
+	}
+}
+
+// ForTaskQueueName builds an exact-match task queue name MutationConstraint.
+func ForTaskQueueName(n string) MutationConstraint {
+	return func(m map[string]interface{}) {
+		m[TaskQueueName.String()] = n
+	}
+}
+
+// ForTaskType builds an exact-match task type MutationConstraint.
+func ForTaskType(t int) MutationConstraint {
+	return func(m map[string]interface{}) {
+		m[TaskType.String()] = t
+	}
+}
+
+// Set assigns the supplied value along with optional constraints to the
+// indicated key. Any other additional value (possibly with other constraints)
+// is overwritten.
+func Set(k Key, v interface{}, constraints ...MutationConstraint) Mutation {
+	return func(m configValueMap) {
+		kstr := strings.ToLower(k.String())
+		cv := constrainedValue{Value: v, Constraints: map[string]interface{}{}}
+		for _, c := range constraints {
+			c(cv.Constraints)
+		}
+		m[kstr] = []*constrainedValue{&cv}
+	}
+}
+
+// Add appends the supplied value and optional constraints to those already
+// present under the indicated key. If there are no values present for the key
+// this is equivalent to Set.
+func Add(k Key, v interface{}, constraints ...MutationConstraint) Mutation {
+	return func(m configValueMap) {
+		kstr := strings.ToLower(k.String())
+		cv := constrainedValue{Value: v, Constraints: map[string]interface{}{}}
+		for _, c := range constraints {
+			c(cv.Constraints)
+		}
+		m[kstr] = append(m[kstr], &cv)
+	}
+}
+
+// Updated applies the provided Mutations to this MutableEphemeralClients
+// underlying non-durable value store.
+func (c *MutableEphemeralClient) Update(mutations ...Mutation) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	current := c.values.Load().(configValueMap)
+	newvals := make(configValueMap, len(current))
+	for k, v := range current {
+		newvals[k] = v
+	}
+	for _, mutate := range mutations {
+		mutate(newvals)
+	}
```

Comment:
this block is trying to make a deep copy, then apply the changes, can you double verify `newvals[k] = v` is also a deep copy?

Reply:
good point. It's shallow (which is wrong)

Reply:
After looking at this some more, I need to v (the slice) but I _don't_ need a deep copy of it. None of the Mutation implementations can modify existing values, they can only assign to the map (which is being copied) or append to the slice (which I will make the change to copy).

This is good because a general-purpose recursive deep-clone of an `interface{}` value is serious business that I'd prefer to avoid.